include/circt/Dialect/LTL/LTLOps.td

```diff
@@ -0,0 +1,312 @@
+// Properties
+//===----------------------------------------------------------------------===//
+
+def NotOp : LTLOp<"not", [Pure]> {
+  let arguments = (ins LTLAnyPropertyType:$input);
+  let results = (outs LTLPropertyType:$result);
+  let assemblyFormat = [{
+    $input attr-dict `:` type($input)
+  }];
+
+  let summary = "A negation of a property.";
+  let description = [{
+    Negates the `$input` property. The resulting property evaluates to true if
+    `$input` evaluates to false, and it evaluates to false if `$input` evaluates
+    to true.
+  }];
+}
+
+def ImplicationOp : LTLOp<"implication", [Pure]> {
+  let arguments = (ins LTLAnySequenceType:$antecedent,
+                       LTLAnyPropertyType:$consequent);
+  let results = (outs LTLPropertyType:$result);
+  let assemblyFormat = [{
+    operands attr-dict `:` type(operands)
+  }];
+
+  let summary = "Only check a property after a sequence matched.";
+  let description = [{
+    Preconditions the checking of the `$consequent` property on the
+    `$antecedent` sequence. In a nutshell, if the `$antecedent` sequence matches
+    at a given point in time, the `$consequent` property is checked starting at
+    the point in time at which the matched sequence ends. The result property of
+    the `ltl.implication` holds if the `$consequent` holds. Conversely, if the
+    `$antecedent` does *not* match at a given point in time, the result property
+    trivially holds. This is conceptually identical to the implication operator
+    →, but with additional temporal semantics.
+  }];
+}
+
+def EventuallyOp : LTLOp<"eventually", [Pure]> {
+  let arguments = (ins LTLAnyPropertyType:$input);
+  let results = (outs LTLPropertyType:$result);
+  let assemblyFormat = [{
+    $input attr-dict `:` type($input)
+  }];
+
+  let summary = "Ensure that a property will hold at some time in the future.";
+  let description = [{
+    Checks that the `$input` property will hold at a future time.
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// Clocking
+//===----------------------------------------------------------------------===//
+
+// Edge behavior enum for always block.  See SV Spec 9.4.2.
+
+/// AtPosEdge triggers on a rise from 0 to 1/X/Z, or X/Z to 1.
+def AtPosEdge: I32EnumAttrCase<"Pos", 0, "posedge">;
+/// AtNegEdge triggers on a drop from 1 to 0/X/Z, or X/Z to 0.
+def AtNegEdge: I32EnumAttrCase<"Neg", 1, "negedge">;
+/// AtEdge is syntactic sugar for AtPosEdge or AtNegEdge.
+def AtEdge   : I32EnumAttrCase<"Both", 2, "edge">;
+
+def ClockEdgeAttr : I32EnumAttr<"ClockEdge", "clock edge",
+                                [AtPosEdge, AtNegEdge, AtEdge]> {
+  let cppNamespace = "circt::ltl";
+}
+
+def ClockOp : LTLOp<"clock", [
+  Pure, InferTypeOpInterface, DeclareOpInterfaceMethods<InferTypeOpInterface>
+]> {
+  let arguments = (ins LTLAnyPropertyType:$input, ClockEdgeAttr:$edge, I1:$clock);
+  let results = (outs LTLSequenceOrPropertyType:$result);
```

Comment:
Is this a new type?  It seems that a clocked property or sequence is the kind of thing that goes to an assert, rather than an unclocked property or sequence.  (e.g. should `clocked` be part of the type?)

Reply:
Having the type represent whether a clock has been provided could be pretty neat. It would make the type system and inference a bit more complicated, but it would also allow us to catch malformed sequences and properties in the IR. Basically, any sequence or property that requires a clock could generated something like an `!ltl.unclocked_sequence` or `!ltl.unclocked_property` (or the type could carry that info), and the `ltl.clock` op would then change this into a `!ltl.sequence` or `!ltl.property`. The asserts could then insist on `!ltl.property` and forbid `!ltl.unclocked_property`.

It's never legal to assert a sequence/property that lacks clocking information. SV just has a ton of weird clock inference rules that can implicitly provide a clock, but that's more of an emission-side optimization than something that should be in the IR.

I'm going to create an issue to extend this in a subsequent PR :+1: